design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,323 @@
+server `B` has determined that server `A` is down, it needs to determine what other server should 
+evaluate the query. At this point,`B` doesn't have any knowledge of the lag information of the other 
+standbys. So, in addition to evaluating the query locally, `B` also sends the request to `C`. `B` 
+then has both its own result and `C`’s result of query evaluation and decides which one is the freshest 
+to include in the response. `B` can make this decision because the lag information is piggybacked 
+with the query evaluation result. The advantages of this approach is that it results in less 
+communication overhead: Lag information is exchanged only when the active is down. Moreover, it is 
+piggy-backed on the request. On the other side, all standbys need to evaluate the same query. Moreover, 
+the communication between `B` and `C` involves large messages as they contain the query result 
+(can be many rows). Finally, latency for a request increases as `B` needs to wait to receive a 
+response from `C` with query result and lag information. Then only can `B` send a response back to 
+the client. 
+
+#### More efficient lag propagation
+Instead of broadcasting lag information, we could also build a gossip protocol to disseminate this 
+information, with more round trips but lower network bandwidth consumption. While we concede that 
+this is an effective and proven technique, debugging such protocols is hard in practice. So, we 
+decided to keep things simple, learn and iterate. Similarly, we could also encode lag information 
+in the regular pull query responses themselves, providing very upto-date lag estimates. However, 
+sending lag information for all local stores in every response will be prohibitively expensive and 
+we would need a more intelligent, selective propagation that only piggybacks a few stores's lag in 
+each response. We may pursue both of these approaches in the future, based on initial experience.
+
+#### Failure detection based on actual request failures
+In this proposal, we have argued for a separate health checking mechanism (i.e separate control plane), 
+while we could have used the pull query requests between servers themselves to gauge whether another 
+server is up or down. But, any scheme like that would require a fallback mechanism that periodically 
+probes other servers anyway to keep the availability information upto date. While we recognize that 
+such an approach could provide much quicker failure detection (and potentially higher number of 
+samples to base failure detection on) and less communication overhead, it also requires significant 
+tuning to handle transient application pauses or other corner cases.
+
+We intend to use the simple heartbeat mechanism proposed here as a baseline implementation, that can 
+be extended to more advanced schemes like these down the line.
+
+## Test plan
+
+We will do unit tests and integration tests with failure scenarios where we cover the cases:
```

Comment:
Should we do scale tests too? We are sending net new RPCs to Kafka to determine lags. With many queries across many instances, this could get expensive right?

Reply:
I read the KLIP before the comments. It may be that no net new RPCs are being issued to Kafka to make this work. Would be good to clarify either way, and then determine if we need to test for impact on Kafka.

Reply:
I would imagine that's a detail for the Kafka KIP, no? I would like to keep the KLIP about ksql facing issues.

On the test for impact on kafka, I think it is subjective on qps that kafka can support for such metadata requests.. and will be highly dependent on the load of a cluster. While we need to be vigilant during deployment and pick sane default, not sure if a test if required.